src/utils/DiscontiguousDataOperations.cu

```diff
@@ -0,0 +1,169 @@
+                                              ValueType *      dst,
+                                              const size_type *discontIds)
+      {
+        const size_type globalThreadId = blockIdx.x * blockDim.x + threadIdx.x;
+        for (size_type i = globalThreadId; i < N * blockSize;
+             i += blockDim.x * gridDim.x)
+          {
+            const size_type blockId      = i / blockSize;
+            const size_type intraBlockId = i - blockId * blockSize;
+            dst[i] = src[discontIds[blockId] * blockSize + intraBlockId];
+          }
+      }
+
+      template <typename ValueType>
+      __global__ void
+      copyToDiscontiguousMemoryDeviceKernel(const size_type  N,
+                                            const size_type  blockSize,
+                                            const ValueType *src,
+                                            ValueType *      dst,
+                                            const size_type *discontIds)
+      {
+        const size_type globalThreadId = blockIdx.x * blockDim.x + threadIdx.x;
+        for (size_type i = globalThreadId; i < N * blockSize;
+             i += blockDim.x * gridDim.x)
+          {
+            const size_type blockId      = i / blockSize;
+            const size_type intraBlockId = i - blockId * blockSize;
+            dst[discontIds[blockId] * blockSize + intraBlockId] = src[i];
+          }
+      }
+
+      template <typename ValueType>
+      __global__ void
+      addToDiscontiguousMemoryDeviceKernel(const size_type  N,
+                                           const size_type  blockSize,
+                                           const ValueType *src,
+                                           ValueType *      dst,
+                                           const size_type *discontIds)
+      {
+        const size_type globalThreadId = blockIdx.x * blockDim.x + threadIdx.x;
+        for (size_type i = globalThreadId; i < N * blockSize;
+             i += blockDim.x * gridDim.x)
+          {
+            const size_type blockId      = i / blockSize;
+            const size_type intraBlockId = i - blockId * blockSize;
+
+            // FIXME: atomicAdd will not work for complex ValueType,
+            // Implement workaround with temporary real and imaginary double
+            // arrays
+            atomicAdd(&dst[discontIds[blockId] * blockSize + intraBlockId],
+                      src[i]);
+          }
+      }
+    } // namespace
+
+    template <typename ValueType>
+    void
+    DiscontiguousDataOperations<ValueType, utils::MemorySpace::DEVICE>::
+      copyFromDiscontiguousMemory(const ValueType *src,
+                                  ValueType *      dst,
+                                  const size_type *discontIds,
+                                  const size_type  N,
+                                  const size_type  blockSize)
+    {
+      copyFromDiscontiguousMemoryDeviceKernel<<<N / dftefe::utils::BLOCK_SIZE +
```

Comment:
`N / dftefe::utils::BLOCK_SIZE` should be `(N *blockSize)/ dftefe::utils::BLOCK_SIZE`

Reply:
Modified